haskell-style.md

```diff
@@ -391,3 +412,134 @@ import qualified Data.List.NonEmpty as NE
 
 import qualified Data.Sequence as Seq
 ```
+
+### Import groups
+
+Imports should be divided into groups:
+
+1. Prelude (when explicit)
```

Comment:
Separating into 3 groups is great, but we generally don't separate even further into qualified and unqualified. 

Also, we generally sort the imports in each group alphabetically.

Reply:
> Also, we generally sort the imports in each group alphabetically.

Yes, this patch includes that rule.

> we generally don't separate even further into qualified and unqualified.

Would folks be against starting to do that? Or is there a strict "let's canonicalize what we currently do (first)" rule at play? Not having this separation makes it harder to maintain the sort, which we agree is important.

Reply:
> Would folks be against starting to do that? 

I personally would. It is common to import container modules qualified and unqualified

```haskell
import Data.Map (Map)
import qualified Data.Map as Map
```

I would rather not have these be split up because of this rule. Also, sometimes you have a diff where all you did in the import list was qualify an import, and with this rule it would move that import to a different line.

I know that `qualified` throws off naive line sorts, but maybe we should have a simple tool that sorts them instead then if we want to automate it (I usually just to a naive sort and then put the qualified modules where they should be, but I also don't remember the last time I've had to do this).

Reply:
Deleted my comment and moving it over here:
> While this looks clean and I've seen it in practice, I don't really see the benefit. A sorted list of imports with prudent use of explicit importing and qualification is easy to traverse. With this you end up having to maintain artificial segments of package types and all new hires must be indoctrinated. Sort your imports is low complexity with high yield.

Reply:
Yeah, I proposed this once before and got pretty much the same response. The OCD in me likes to separated qualified vs. non-qualified, but the obsequious sycophant just wants everyone to be happy. Really glad we're finally writing this down somewhere.

Reply:
How about this simple script?

```
#!/usr/bin/env stack
{-# LANGUAGE OverloadedStrings #-}

import Data.List (sortBy)
import Data.Ord (comparing)
import Data.Text as T
import Data.Text.IO as T
import System.IO

main =
  T.hGetContents stdin
    >>= traverse (T.hPutStrLn stdout) . sortBy (comparing unprefixImport) . T.lines

unprefixImport x
  | T.null x = x
  | otherwise = T.replace " " ""  $ T.replace "qualified" "" $ T.replace "import" "" x
```

Reply:
Sure! but...

```hs
T.interact $ T.unlines . sortBy (comparing unprefixImport) . T.lines
```

Reply:
Alternative script:
```
#!/usr/bin/env bash

ImportRex='\(import\s*\)\(qualified\s*\)\{0,1\}'
ToSortRex="s/^$ImportRex\\(.*\\)$/\\3:::\\1\\2/g"
FromSortRex="s/^\([^:::]*\):::$ImportRex$/\\2\\3\\1/g"

cat | sed "$ToSortRex" | sort | sed "$FromSortRex"
```